fw/http_parser.c

```diff
@@ -3753,6 +3730,39 @@ tfw_http_init_parser_req(TfwHttpReq *req)
 	hbh_hdrs->spec = 0x1 << TFW_HTTP_HDR_CONNECTION;
 }
 
+int
+tfw_http_parse_req_on_headers_done(TfwHttpReq *req)
+{
+#define IS_BODYLESS_METHOD(meth)						\
+	((meth) == TFW_HTTP_METH_GET || (meth) == TFW_HTTP_METH_HEAD		\
+	 || (meth) == TFW_HTTP_METH_DELETE || (meth) == TFW_HTTP_METH_TRACE)
+
+	TfwStr *tbl = req->h_tbl->tbl;
+
+	/* According to RFC 7231 4.3.* a payload within GET, HEAD,
+	 * DELETE, TRACE and CONNECT requests has no defined semantics
+	 * and implementations can reject it. We do this respecting overrides.
+	 */
+	if (TFW_STR_NOT_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_LENGTH])
+	    || TFW_STR_NOT_EMPTY(&tbl[TFW_HTTP_HDR_CONTENT_TYPE]))
+	{
+		/* Method override either honored or request message
+		 * with method override header dropped later in processing */
+		if ((req->method_override
+			&& IS_BODYLESS_METHOD(req->method_override))
+		    || IS_BODYLESS_METHOD(req->method)) {
+				T_WARN("%s: Content-Length or Content-Type"
+				       " not allowed to be used with such"
+				       " (overridden) method\n", __func__);
+				return T_DROP;
+		}
+	}
+
+	return T_OK;
```

Comment:
I'm worried about the return value. Maybe need to use T_POSTPONE?

Reply:
We use the `T_*` return value constants to make the processing logic more explicit - what should we do next: read another chunk, pass or block the current chunk. In this function I think basic true/false would be fine as well as `T_OK` or `T_DROP`